src/java/org/apache/cassandra/service/SSTablesGlobalTracker.java

```diff
@@ -0,0 +1,284 @@
+                    if (newValue == 0)
+                        sstablesForOtherVersions.remove(version);
+                    else
+                        sstablesForOtherVersions.put(version, newValue);
+                }
+            }
+
+            if (triggerUpdate)
+                versionsInUse = computeVersionsInUse(sstablesForCurrentVersion, currentVersion, sstablesForOtherVersions);
+        }
+        return triggerUpdate;
+    }
+
+    private static ImmutableSet<VersionAndType> computeVersionsInUse(int sstablesForCurrentVersion, VersionAndType currentVersion, Map<VersionAndType, Integer> sstablesForOtherVersions)
+    {
+        ImmutableSet.Builder<VersionAndType> builder = ImmutableSet.builder();
+        if (sstablesForCurrentVersion > 0)
+            builder.add(currentVersion);
+        builder.addAll(sstablesForOtherVersions.keySet());
+        return builder.build();
+    }
+
+    private static int sanitizeSSTablesCount(int sstableCount, VersionAndType version)
+    {
+        if (sstableCount >= 0)
+            return sstableCount;
+
+        /*
+         This shouldn't happen and indicate a bug either in the tracking of this class, or on the passed notification.
+         That said, it's not worth bringing the node down, so we log the problem but otherwise "correct" it.
+        */
+        noSpamLogger.error("Invalid state while handling sstables change notification: the number of sstables for " +
+                           "version {} was computed to {}. This indicate a bug and please report it, but it should " +
+                           "not have adverse consequences.", version, sstableCount, new RuntimeException());
+        return 0;
+    }
+
+    private static Iterable<Descriptor> addedSSTables(INotification notification)
+    {
+        if (notification instanceof SSTableAddedNotification)
+            return Iterables.transform(((SSTableAddedNotification)notification).added, s -> s.descriptor);
+        else if (notification instanceof SSTableListChangedNotification)
+            return Iterables.transform(((SSTableListChangedNotification)notification).added, s -> s.descriptor);
+        else if (notification instanceof SSTableAddedDuringInitializationNotification)
+            return Iterables.transform(((SSTableAddedDuringInitializationNotification)notification).added, s -> s.descriptor);
+        else
+            return Collections.emptyList();
+    }
+
+    private static Iterable<Descriptor> removedSSTables(INotification notification)
+    {
+        if (notification instanceof SSTableDeletingNotification)
+            return Collections.singletonList(((SSTableDeletingNotification)notification).deleting.descriptor);
+        else if (notification instanceof SSTableListChangedNotification)
+            return Iterables.transform(((SSTableListChangedNotification)notification).removed, s -> s.descriptor);
+        else
+            return Collections.emptyList();
+    }
```

Comment:
The `else` are unnecessary as all the statements end with a `return`.

To be honest, I would be in favor of introducing an `SSTablesChangedNotification` as a super interface of `SSTableAddedNotification`, `SSTableListChangedNotification`, `SSTableAddedDuringInitializationNotification `and `SSTableDeletingNotification`.

```
/**
 * Notification sent when SSTables are added to and/or removed from their {@link org.apache.cassandra.db.ColumnFamilyStore}.
 */
public interface SSTablesChangedNotification extends INotification
{
    /**
     * Checks if the notification is due to SSTables being initialized. 
     *
     * @return {@code true} if the notification is due to SSTables being initialized, {@code false} otherwise.
     */
    public default boolean initialization()
    {
        return false;
    }

    /**
     * Returns the SSTables that have been added to the ColumnFamilyStore.
     *
     * @return the SSTables that have been added to the ColumnFamilyStore.
     */
    public default Iterable<SSTableReader> addedSSTables()
    {
        return Collections.emptyList();
    }

    /**
     * Returns the SSTables that have been removed from the ColumnFamilyStore.
     *
     * @return the SSTables that have been removed from the ColumnFamilyStore.
     */
    public default Iterable<SSTableReader> removedSSTables()
    {
        return Collections.emptyList();
    }
}
```

Reply:
Oh I saw the `else` and forgot to remove it :( Thanks!
About the interface - I find it a great idea but I feel a bit uncomfortable to do more refactoring around the notifications at this point in time just before RC. Let me know if you think I am too cautious but otherwise I would refrain for now from more changes, even if I totally agree and see the clarity in your proposal.

Reply:
I am not pushing for it :-)